File: task4/generic_collection/src/main/java/collection/MyList.java
package collection;

import java.util.Comparator;
import java.util.Optional;


public interface MyList<T> {
    boolean add(T object);
    boolean add(int index, T object);
    Optional<T> get(int index);
    Optional<T> remove(int index);
    boolean remove(T object);
    int size();
    Optional<T> min(Comparator<T> comparator);
    Optional<T> max(Comparator<T> comparator);
    Optional<T> sum(ToNumberFunction<? super T> mapper);
    Optional<T> product(ToNumberFunction<? super T> mapper);
    int indexOf(T object);
    boolean contains(T object);
    void sortBubble(Comparator<T> comparator);
    void sortInsertion(Comparator<T> comparator);
    void sortSelection(Comparator<T> comparator);
    Optional<T> set(int index, T object);








}

File: task7/MyArrayList/src/main/kotlin/collection/MyList.kt
package collection

interface MyList <T>{
    fun size(): Int
    fun add(obj: T): Boolean
    fun add(index: Int, obj: T): Boolean
    fun get(index: Int): T?
    fun remove(index: Int): T?
    fun remove(obj: T): Boolean
    fun min(comparator: Comparator<in T>): T?
    fun max(comparator: Comparator<in T>): T?
    fun indexOf(obj: T): Int
    fun contains(obj: T): Boolean
    fun set(index: Int, obj: T): T?
    fun sortBubble(comparator: Comparator<in T>)
    fun sortInsertion(comparator: Comparator<in T>)
    fun sortSelection(comparator: Comparator<in T>)
    fun sum(mapper: ToNumberFunction<in T>?): T?
    fun product(mapper: ToNumberFunction<in T>?): T?
}